src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -211,8 +213,11 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 self.const_usize(self.layout_of(tp_ty).align.pref.bytes())
             }
             "type_name" => {
-                let tp_ty = substs.type_at(0);
-                let ty_name = self.tcx.type_name(tp_ty);
+                let gid = GlobalId {
+                    instance,
+                    promoted: None,
+                };
+                let ty_name = self.tcx.const_eval(ty::ParamEnv::reveal_all().and(gid)).unwrap();
```

Comment:
Can you leave a comment saying all zero-argument intrinsics should move to this approach?

Reply:
Wow so the implementation in `librustc_mir` also gets used at run-time with this? Scary.^^